### ISSUE TITLE
Feature Request: Diagram 'button' to execute a script

### ISSUE THREAD

**Issue by GuyWicks** [archimatetool/archi-scripting-plugin#39]
TL;DR - Allow 'scripts' objects to be dropped onto a view/diagram, which creates a 'button' that when clicked, executes the script.

It could be useful to have 'buttons' (or a series of buttons) that can be linked to scripts to allow, for example, different 'colour-view' scripts to be executed - a use case that I tend to use a lot.

I would imaging that:

- The button would be a 'drawing only' object. It would not be a part of the Archimate concepts or relations
- You'd probably would not want it printed/exported to PDF & PNG etc
- It would have properties for 'name', 'script', some styling 
- It could even have 'properties' that could be passed to the script to provide a form of individual context (i.e. different buttons run the same script with different parameters)
- The script could be 'dragged and dropped' onto the canvas (like views can be) to create the button.

**Comment by Phillipus** [comment]
Might be a problem sharing that model file with someone who doesn't have the jArchi plugin?

**Comment by GuyWicks** [comment]
True.  Would it be reasonable to expect that Archi, in the event of finding a button element but without the corresponding 'plug-in' would either:

a - become disabled (i.e. grayed out with no click-ability)
b - emit a message ( 'No corresponding plug-in found for 'insert this button title here' )

Also, there would need to be some 'backwards compatibility' thought in place for old versions of Archi.  IIRC, Archi strips out any 'foreign' metadata from the file when it is saved, so if a new model was opened in an old version, it would clobber the buttons - unless the 'metadata version' number in the file somehow could warn against that. 

Finally, how would this new 'button' work in the collaboration components (thinking in particular the database ones...   hmmm - might be making this sound harder than I would like it to be! :)

The thing about the Archi, is that it's users are predominantly architects (SA in my case) so we're always trying to SOLVE problems!

**Comment by Phillipus** [comment]
Archi code knows nothing about down-stream plugins, so it can't decide "If this is something from plugin x".

TBH, this is not something that I'm going to implement in Archi itself. However, a more generic solution might be along the lines of "generate an event when x happens on an object". The event would be part of a generic framework where a plugin could respond. So, for example, jArchi could register an event listener on a click on visual object to run a script.

> Archi strips out any 'foreign' metadata from the file when it is saved

This will change going forward for new features.

**Comment by GuyWicks** [comment]
Understood. I saw the FR about event registrations and thought this would be a good use-case for it.  

The "drag-and-drop" of the script onto the view to would still be a nice UX feature...
